lib/ramble/ramble/keywords.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2022-2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+# <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+
+from enum import Enum
+
+import llnl.util.tty as tty
+
+import ramble.error
+
+
+class Keywords(object):
+    """Class to represent ramble known keywords.
+
+    Each keyword can have a type. Valid types are in the class level
+    'key_types' variable as an enum.
+
+    A map of keys, and their types is provided in the class level 'keys'
+    variable. This enforces that each keyword can only have a single type.
+
+    Current key types are:
+    - Reserved: Ramble defines these, and a user should not be allowed to define them
+    - Optional: Ramble can function with a definition from the user but it isn't required
+    - Required: Ramble requires a definition for these. Ramble will try to set sensible defaults,
+                but it might not be possible always.
+
+    """
+
+    key_types = Enum('types', ['reserved', 'optional', 'required'])
+
+    keys = {
+        'application_name': key_types.reserved,
+        'application_run_dir': key_types.reserved,
+        'application_input_dir': key_types.reserved,
+        'workload_name': key_types.reserved,
+        'workload_run_dir': key_types.reserved,
+        'workload_input_dir': key_types.reserved,
+        'experiment_name': key_types.reserved,
+        'experiment_run_dir': key_types.reserved,
+        'log_dir': key_types.reserved,
+        'log_file': key_types.reserved,
+        'err_file': key_types.reserved,
+        'command': key_types.reserved,
+        'spack_env': key_types.reserved,
+
+        'spec_name': key_types.optional,
+
+        'n_ranks': key_types.required,
+        'n_nodes': key_types.required,
+        'processes_per_node': key_types.required,
+        'n_threads': key_types.required,
+        'batch_submit': key_types.required,
+        'mpi_command': key_types.required,
+    }
+
+    def __init__(self):
+        # Define class attributes for all of the keys
+        for key in self.keys.keys():
+            setattr(self, key, key)
+
+    def is_valid(self, key):
```

Comment:
If we use a more traditional enum then can this class be static? Right now we use a lot of `self` but don't really hold state. That would let us do something more like:

```
Keywords.is_reservered(...)
```
Rather than the current
```
keywords = Keywords()
keywords.is_reservered(...)
```

Reply:
Agreed.

I can make most of these static methods.

I don't actually think I can use traditional enums after looking at this more, at least not without making the code more complex.

One option to help with this `Keywords.is_reserved(...)` aspect is to create singleton instance of the class (as is done in config), and change the references to that instead of instantiating the class multiple times.

Reply:
FWIW, I made these class methods. But then added the singleton of the class so we don't need to refer to it by instantiating it anymore.